xaitk_saliency/impls/vis_sal_similarity/similarityscoring.py

```diff
@@ -0,0 +1,98 @@
+    `PerturbImage` implementation.
+
+    The perturbation masks used by the following implementation are
+    expected to be of type integer. Masks containing values of type
+    float are rounded to the nearest value and binarized
+    with value 1 replacing values greater than or equal to half of
+    the maximum value in mask after rounding while 0 replaces the rest.
+
+    param proximity_metric: The type of comparision metric used
+        to determine proximity in feature space. The type of comparision
+        metric supported is restricted by scipy's cdist() function. The
+        following metrics are supported in scipy.
+
+        ‘braycurtis’, ‘canberra’, ‘chebyshev’, ‘cityblock’, ‘correlation’,
+        ‘cosine’, ‘dice’, ‘euclidean’, ‘hamming’, ‘jaccard’, ‘jensenshannon’,
+        ‘kulsinski’, ‘mahalanobis’, ‘matching’, ‘minkowski’, ‘rogerstanimoto’,
+        ‘russellrao’, ‘seuclidean’, ‘sokalmichener’, ‘sokalsneath’,
+        ‘sqeuclidean’, ‘wminkowski’, ‘yule’.
+    """
+
+    def __init__(
+        self,
+        proximity_metric: str = 'euclidean'
+    ):
+
+        try:
+            # Attempting to use chosen comparision metric
+            cdist([[1], [1]], [[1], [1]], proximity_metric)
+            self.proximity_metric: str = proximity_metric
+        except ValueError:
+            raise ValueError("Chosen comparision metric not supported or",
+                             "may not be available in scipy")
+
+    def generate(
+        self,
+        ref_descr_1: np.ndarray,
+        ref_descr_2: np.ndarray,
+        perturbed_descrs: np.ndarray,
+        perturbed_masks: np.ndarray,
+    ) -> np.ndarray:
+
+        if len(perturbed_descrs) != len(perturbed_masks):
+            raise ValueError("Number of perturbation masks and respective",
+                             "feature vector do not match.")
+
+        if len(ref_descr_1) != len(ref_descr_2):
+            raise ValueError("Length of feature vector between",
+                             "two images do not match.")
+
+        # Computing original proximity between image1 and image2 feature vectors.
+        original_proximity = cdist(
+            ref_descr_1.reshape(1, -1),
+            ref_descr_2.reshape(1, -1),
+            metric=self.proximity_metric
+        )
+
+        # Computing proximity between original image1 and perturbed image2 feature vectors.
+        perturbed_proximity = cdist(
+            ref_descr_1.reshape(1, -1),
+            perturbed_descrs,
+            metric=self.proximity_metric
+        )[0]
+
+        # Iterating through each distance and compare it with
+        # its perturbed twin
+        diff = perturbed_proximity - original_proximity
+
+        diff = np.transpose(np.clip(diff, 0, None))
+        # Weighting perturbed regions with respective difference in confidence
+        sal = weight_regions_by_scalar(diff, perturbed_masks)
+
+        # Converting nan values to zero.
+        sal = np.nan_to_num(sal)
```

Comment:
I assume this is the line that may be removed with the update to the `weight_regions_by_scalar` function?

Reply:
This will probably also require a separate PR for the OcclusionScoring method as well?

https://github.com/XAITK/xaitk-saliency/blob/master/xaitk_saliency/impls/vis_sal_classifier/occlusion_scoring.py#L48

Reply:
Given the scale of where we are, I won't complain with just removing that line in `occlusion_scoring.py` in this PR.

Reply:
Sounds good, I removed the nan_to_num conversion in both occlusion_scoring.py and similarityscoring.py